### PR TITLE
[otbn] Speed up the ISS

### PR DESCRIPTION
**Pull request by rswarbrick**
This series speeds up the ISS by roughly a factor of 2 on a silly loop that runs `nop` a million times. The speed-ups will be less for programs that actually do some work.

@felixmiller: I know you've been struggling with the slow simulator. Could you give this a spin to see if it improves things?

### CODE REVIEW

**Comment by felixmiller** [comment]
@rswarbrick: In [here](https://github.com/felixmiller/opentitan/tree/otbn-rsa-test/sw/otbn/code-snippets) is now self-contained assembly for [encrypt](https://github.com/felixmiller/opentitan/blob/otbn-rsa-test/sw/otbn/code-snippets/rsa_enc_test.S) and [decrypt](https://github.com/felixmiller/opentitan/blob/otbn-rsa-test/sw/otbn/code-snippets/rsa_dec_test.S) with RSA 1024. This should reproduce the issue in the ISS.

**Comment by rswarbrick** [comment]
Thanks! The problem was in the disassembly memoization changes. They did look like this:
```python
    ...
    if state.stalled:
        insn = None
        ...
    else:
        insn = something
        ...
    ...
    state.mutate_yourself()
    if not state.stalled:
        insn.something
```

The problem is that `state.mutate_yourself()` cleared the `stalled` flag, so the second check didn't match the first.

Anyway, this is all fixed now (by taking a copy of the flag).

**Review by felixmiller — APPROVED** [review]
Thanks, LGTM. Now at 31s for RSA 2048 decrypt without any problems

**Review by imphil — APPROVED** [review]
LGTM. I'm not a fan of the "[otbn] Accumulate changes "imperatively" in ISS " commit, the API looks strange to me. Since it's only a minimal improvement on a micro-benchmark, I'd remove it.

File: hw/ip/otbn/dv/otbnsim/sim/isa.py [hunk 1]
```diff
@@ -73,12 +71,24 @@ class OTBNInsn:
...
+
     def execute(self, state: OTBNState) -> None:
         raise NotImplementedError('OTBNInsn.execute')
 
     def disassemble(self, pc: int) -> str:
         '''Generate an assembly listing for this instruction'''
-        return self.insn.disassemble(self.op_vals, 12)
+        if self._disasm is not None:
+            old_pc, old_disasm = self._disasm
+            if pc == old_pc:
```
[the remark this answers is not in the document]
Comment (rswarbrick) [reply hw/ip/otbn/dv/otbnsim/sim/isa.py:86]: Good point. For some reason, I had the idea that it might change, but that's nonsense. I'll ditch it!

**Comment by rswarbrick** [comment]
> I'm not a fan of the "[otbn] Accumulate changes "imperatively" in ISS " commit, the API looks strange to me. Since it's only a minimal improvement on a micro-benchmark, I'd remove it.

Yeah, I was a bit conflicted. 2% is not nothing, but the ISS is pretty slow in general (a result of writing it in Python); maybe it's not worth the clarity trade-off. Will drop.

**Pull request merged**